### PR TITLE
Fixed #19580 -- Made nullable relations work consistently

### PR DESCRIPTION
**Pull request by Rajesh-Veeranki**
Rebased the previous PR https://github.com/django/django/pull/7098/ with the master here. Modified the release text. The tests fail when I remove any newly added line. Please review and suggest any changes.

### CODE REVIEW

**Comment by timgraham** [comment]
This doesn't address my comment from the other PR:  "Looking at the rest of it, I'm not sure what happened to much of Anssi's original patch (master...akaariai:ticket_17541). It seems that only the many-to-many changes survived the rebase?"

File: docs/releases/1.11.2.txt [hunk 1]
```diff
@@ -39,3 +39,32 @@ Bugfixes
 
 * Fixed ``MultipleObjectMixin.paginate_queryset()`` crash on Python 2 if the
   ``InvalidPage`` message contains non-ASCII (:ticket:`28204`).
+
+.. _backwards-incompatible-1.11.2:
+
+Backwards incompatible changes in 1.11.2
```
Comment (timgraham) [inline docs/releases/1.11.2.txt]: This should target Django 2.0.

**Comment by Rajesh-Veeranki** [comment]
You are right, the RelatedManager changes went missing. I added them back now. But more tests are required. I will add them in a few days.

**Comment by Rajesh-Veeranki** [comment]
I have added changes and tests for RelatedManager. Please review them.

While writing the test cases, I came across this. When `Choice` is related to `Poll` by poll.question, i.e.


    class Poll(models.Model):
        question = models.CharField(max_length=200, unique=True, null=True)

        def __str__(self):
            return "Q: %s " % self.question


    class Choice(models.Model):
        poll = models.ForeignKey(Poll, models.SET_NULL, to_field='question', null=True)
        choice = models.CharField(max_length=200)

        def __str__(self):
            return "Choice: %s in poll %s" % (self.choice, self.poll)

Is this piece of code valid?
 
`p = Poll.objects.create(question=None)`
`c = Choice.objects.create(poll=p, choice='A')`
`p.choice_set.count() == 0`
The current django code evaluates them without error

File: tests/m2m_through_regress/tests.py [hunk 2]
```diff
@@ -229,6 +230,15 @@ def test_m2m_relations_unusable_on_null_pk_obj(self):
         with self.assertRaisesMessage(ValueError, msg):
             Car(make='Ford').drivers.all()
 
+    def test_add_null(self):
+        nullcar = Car.objects.create(make=None)
+        with self.assertRaisesMessage(
+            ValueError,
+            '"<Car: None>" needs to have a value for field "make" '
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: Use the `msg  = ` style of the first test.

File: tests/many_to_many/tests.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ def setUp(self):
 
     def test_add(self):
         # Create an Article.
-        a5 = Article(headline='Django lets you reate Web apps easily')
+        a5 = Article(headline='Django lets you create Web apps easily')
```
Comment (timgraham) [inline tests/many_to_many/tests.py]: Make this unrelated fix in a separate pull request.

File: tests/many_to_one_null/models.py [hunk 4]
```diff
@@ -32,3 +32,18 @@ class Car(models.Model):
 
 class Driver(models.Model):
     car = models.ForeignKey(Car, models.SET_NULL, to_field='make', null=True, related_name='drivers')
+
+
+class Poll(models.Model):
```
Comment (timgraham) [inline tests/many_to_one_null/models.py]: Is there an issue reusing the `Car`/`Driver` (or some other) models? We prefer not to add new models unless absolutely necessary since that increases complexity and the time needed to run the tests.

File: tests/many_to_one_null/tests.py [hunk 5]
```diff
@@ -134,3 +134,25 @@ def test_related_null_to_field(self):
         self.assertIs(d1.car, None)
         with self.assertNumQueries(0):
             self.assertEqual(list(c1.drivers.all()), [])
+
+    def test_add_to_null_related_field(self):
+        p = Poll.objects.create(question=None)
+        c = Choice.objects.create(choice='B')
+
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/many_to_one_null/tests.py]: Use `assertRaiseMessage` as in the other tests.

File: tests/null_queries/tests.py [hunk 6]
```diff
@@ -39,9 +39,9 @@ def test_none_as_null(self):
         with self.assertRaises(ValueError):
             Choice.objects.filter(id__gt=None)
 
-        # Related managers use __exact=None implicitly if the object hasn't been saved.
         p2 = Poll(question="How?")
-        self.assertEqual(repr(p2.choice_set.all()), '<QuerySet []>')
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/null_queries/tests.py]: Use `assertRaiseMessage` as in the other tests (maybe this is better with the other added tests though).

File: docs/releases/2.0.txt [hunk 7]
```diff
@@ -380,6 +380,28 @@ raises an exception and should be replaced with::
...
+The behavior of nullable related ``QuerySet`` has changed.
+
+Nullable related ``QuerySet`` means a ``QuerySet`` for reverse relation
+to a field which can be set to null.
+
+In 1.11.x querying a nullable many-to-many relation was an error. For foreign
+key relations there was no error. To make the nullable relations consistent
+the following changes were made:
+
+* Nullable many-to-many and foreign key relations will return an empty
```
Comment (timgraham) [inline docs/releases/2.0.txt]: Some code examples to illustrate these points would be useful.

File: django/db/models/fields/related_descriptors.py [hunk 8]
```diff
@@ -570,7 +576,15 @@ def get_prefetch_queryset(self, instances, queryset=None):
             cache_name = self.field.related_query_name()
             return queryset, rel_obj_attr, instance_attr, False, cache_name
 
+        def _check_fk_val(self):
+            for field in self.field.foreign_related_fields:
+                if getattr(self.instance, field.attname) is None:
+                    raise ValueError('"%r" needs to have a value for field "%s" before '
```
Comment (timgraham) [inline django/db/models/fields/related_descriptors.py]: Use same hanging indent style as line 516.

File: docs/releases/2.0.txt (hunk 7 above)
Comment (Rajesh-Veeranki) [inline docs/releases/2.0.txt]: Added them now. Please have a look.

File: tests/many_to_one_null/models.py (hunk 4 above)
Comment (Rajesh-Veeranki) [inline tests/many_to_one_null/models.py]: My bad! I could reuse the models. Fixed it now.

**Comment by Rajesh-Veeranki** [comment]
Hi @carltongibson 
Thanks for the review. Apologies for my delayed response.
Please find my reply to the corresponding comments below

1. I added `_check_has_fk_val` to `_add_items`, `_remove_items`, `clear` only because all the other functions gets called through them. So it's deliberate.
2. Adjusted now
3. I didn't get this comment. Can you please elaborate. Do you want me to move all the tests to null_queries/tests.py? 
4. Please give me the text for the docs. I can update my PR with it.

**Comment by carltongibson** [comment]
HI @Rajesh-Veeranki. 

Please: 

* Clear up the failure to the docs build. 
* Move the relevant check from `null_queries/tests.py` to `many_to_one_null/tests.py` so they're all grouped, [as per Tim's comment](https://github.com/django/django/pull/8570#discussion_r121480284)
* Squash and rebase, making sure [the commit message uses the past tense](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines).

If you then uncheck "Patch need improvement" on the ticket, I will have another look, and adjust the wording on the docs. 

Thanks for your effort!

File: tests/many_to_one_null/models.py (hunk 4 above)
Comment (carltongibson) [reply tests/many_to_one_null/models.py]: `Poll` and `Choice` can still be removed. The test using them just requires an unsaved model, so `Car()` triggers the error.

**Comment by Rajesh-Veeranki** [comment]
Hi @carltongibson 
Thanks for the review. I'll work on your suggestions soon and reach out in case I've any doubt.

File: tests/many_to_one_null/models.py (hunk 4 above)
Comment (auvipy) [reply tests/many_to_one_null/models.py]: should we remove all the Poll and Choice models and related codes?

**Comment by carltongibson** [comment]
Hi @Rajesh-Veeranki. Are you up for finishing this PR? (If not I can de-assign it and I'm sure someone would pick it up to get it across the line.)

**Comment by LuRsT** [comment]
Gentle reminder @Rajesh-Veeranki, are you planning on picking this up again?

**Comment by Rajesh-Veeranki** [comment]
Hi @LuRsT , I am NOT planning to pick this up anytime soon. So feel free to deassign. Thanks
